src/lib/core/GroupedCallbackList.h

```diff
@@ -0,0 +1,213 @@
+void TaggedDequeueGroup(Cancelable * cancelable);
+void EnqueueWithGroup(Cancelable * cancelable, Cancelable *& group, Cancelable * anchor, void (*cancelFn)(Cancelable *));
+void LinkGroup(Cancelable * prev, Cancelable * next);
+inline Cancelable * ClearCancelable(Cancelable * cancelable);
+} // namespace detail
+
+/**
+ * A GroupedCallbackList is similar to a CallbackDeque, but manages groups of
+ * related callbacks (possibly of different types) in a single list.
+ *
+ * Groups of callbacks are enqueued and dequeued (or cancelled) as a unit.
+ * Within a group each callback is optional (i.e. can be null), however attempting
+ * to enqueuea group where all callbacks are null has no effect.
+ */
+template <typename... T>
+class GroupedCallbackList : protected Cancelable
+{
+    static_assert(sizeof...(T) > 0);
+
+public:
+    bool IsEmpty() { return mNext == this; }
+
+    /**
+     * Enqueues the specified group of callbacks, any of which may be null.
+     */
+    void Enqueue(Callback<T> *... callback) { Enqueue(std::index_sequence_for<T...>{}, callback...); }
+
+    /**
+     * If the list is non-empty, populates the reference arguments with the first
+     * group of callbacks and returns true. Returns false if the list is empty.
+     */
+    bool Peek(Callback<T> *&... callback) const { return Peek(std::index_sequence_for<T...>{}, callback...); }
+
+    /**
+     * Like Peek(), but additionally removes the first group of callbacks from the list.
+     */
+    bool Take(Callback<T> *&... callback)
+    {
+        VerifyOrReturnValue(Peek(callback...), false);
+        mNext->Cancel();
+        return true;
+    }
+
+    /**
+     * Moves all elements of the source list into this list, leaving the source list empty.
+     */
+    void EnqueueTakeAll(GroupedCallbackList & source)
+    {
+        VerifyOrReturn(!source.IsEmpty() && this != &source);
+        detail::LinkGroup(mPrev, source.mNext);
+        source.mPrev->mNext = this;
+        mPrev               = source.mPrev;
+
+        source.mPrev = source.mNext = &source;
+    }
+
+private:
+    // The grouped list structure is similar to the doubly linked list used by
+    // CallbackDeque, however we have the additional requirement of representing
+    // node grouping. Due to the requirement so support sparse groups (one or more
+    // callbacks may not be present in a particular group) we cannot rely on a
+    // fixed group size. This problem is solved by having the "prev" pointer for
+    // all nodes in a group point to the node before the group, as illustrated below:
+    //
+    //              |== Grouped Nodes ==|
+    //
+    //   +---+      +---+   +---+   +---+      +---+
+    //   |   |----->|   |-->|   |-->|   |----->|   |
+    //   |   |      |   |   |   |   |   |      |   |
+    //   |   |<-----|   |  -|   |  -|   |<-----|   |
+    //   +---+  \   +---+ / +---+ / +---+      +---+
+    //           \_______/       /
+    //            \_____________/
+    //
```

Comment:
So in this picture, what happens when the grouped nodes are canceled?  I guess the cancel func this is using will remove the whole group if any member of the group gets canceled?

Reply:
Correct, groups are enqueued and dequeued / cancelled as a unit. The cancel func is `DequeueGroup`.